File: app/src/main/java/com/dertyp7214/appstore/uiTools/ClipRectAnimation.java
/*
 * Copyright (c) 2018.
 * Created by Josua Lengwenath
 */

package com.dertyp7214.appstore.uiTools;

import android.graphics.Rect;
import android.graphics.RectF;
import android.view.animation.Animation;
import android.view.animation.Transformation;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * An animation that controls the clip of an object. See the
 * {@link android.view.animation full package} description for details and
 * sample code.
 */
public class ClipRectAnimation extends Animation {
    private RectF mFromRect = new RectF();
    private RectF mToRect = new RectF();

    private Rect mResolvedFrom = new Rect();
    private Rect mResolvedTo = new Rect();
    private Method mSetClipRect;

    /**
     * Constructor to use when building a ClipRectAnimation from code
     *
     * @param fromClip the clip rect to animate from
     * @param toClip the clip rect to animate to
     */
    public ClipRectAnimation(RectF fromClip, RectF toClip) {
        if (fromClip == null || toClip == null) {
            throw new RuntimeException("Expected non-null animation clip rects");
        }
        mFromRect.set(fromClip);
        mToRect.set(toClip);
    }

    /**
     * Constructor to use when building a ClipRectAnimation from code
     */
    public ClipRectAnimation(float fromL, float fromT, float fromR, float fromB,
                             float toL, float toT, float toR, float toB) {
        mFromRect.set(fromL, fromT, fromR, fromB);
        mToRect.set(toL, toT, toR, toB);
    }

    @Override
    public void initialize(int width, int height, int parentWidth, int parentHeight) {
        super.initialize(width, height, parentWidth, parentHeight);
        mResolvedFrom.left = (int) resolveSize(RELATIVE_TO_SELF, mFromRect.left, width, parentWidth);
        mResolvedFrom.top = (int) resolveSize(RELATIVE_TO_SELF, mFromRect.top, height, parentHeight);
        mResolvedFrom.right = (int) resolveSize(RELATIVE_TO_SELF, mFromRect.right, width, parentWidth);
        mResolvedFrom.bottom = (int) resolveSize(RELATIVE_TO_SELF, mFromRect.bottom, height, parentHeight);
        mResolvedTo.left = (int) resolveSize(RELATIVE_TO_SELF, mToRect.left, width, parentWidth);
        mResolvedTo.top = (int) resolveSize(RELATIVE_TO_SELF, mToRect.top, height, parentHeight);
        mResolvedTo.right = (int) resolveSize(RELATIVE_TO_SELF, mToRect.right, width, parentWidth);
        mResolvedTo.bottom = (int) resolveSize(RELATIVE_TO_SELF, mToRect.bottom, height, parentHeight);
    }

    @Override
    protected void applyTransformation(float it, Transformation tr) {
        int l = mResolvedFrom.left + (int) ((mResolvedTo.left - mResolvedFrom.left) * it);
        int t = mResolvedFrom.top + (int) ((mResolvedTo.top - mResolvedFrom.top) * it);
        int r = mResolvedFrom.right + (int) ((mResolvedTo.right - mResolvedFrom.right) * it);
        int b = mResolvedFrom.bottom + (int) ((mResolvedTo.bottom - mResolvedFrom.bottom) * it);

        try {
            mSetClipRect.invoke(tr, l, t, r, b);
        } catch (IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
    }

    @Override
    public boolean willChangeTransformationMatrix() {
        return false;
    }
}

File: app/src/main/java/com/dertyp7214/appstore/uiTools/ClipRectAnimation.kt
/*
 * Copyright (c) 2018.
 * Created by Josua Lengwenath
 */

package com.dertyp7214.appstore.uiTools

import android.graphics.Rect
import android.graphics.RectF
import android.view.animation.Animation
import android.view.animation.Transformation

import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

class ClipRectAnimation : Animation {
    private val mFromRect = RectF()
    private val mToRect = RectF()

    private val mResolvedFrom = Rect()
    private val mResolvedTo = Rect()
    private val mSetClipRect: Method? = null

    constructor(fromClip: RectF?, toClip: RectF?) {
        if (fromClip == null || toClip == null) {
            throw RuntimeException("Expected non-null animation clip rects")
        }
        mFromRect.set(fromClip)
        mToRect.set(toClip)
    }

    constructor(fromL: Float, fromT: Float, fromR: Float, fromB: Float,
                toL: Float, toT: Float, toR: Float, toB: Float) {
        mFromRect.set(fromL, fromT, fromR, fromB)
        mToRect.set(toL, toT, toR, toB)
    }

    override fun initialize(width: Int, height: Int, parentWidth: Int, parentHeight: Int) {
        super.initialize(width, height, parentWidth, parentHeight)
        mResolvedFrom.left = resolveSize(Animation.RELATIVE_TO_SELF, mFromRect.left, width, parentWidth).toInt()
        mResolvedFrom.top = resolveSize(Animation.RELATIVE_TO_SELF, mFromRect.top, height, parentHeight).toInt()
        mResolvedFrom.right = resolveSize(Animation.RELATIVE_TO_SELF, mFromRect.right, width, parentWidth).toInt()
        mResolvedFrom.bottom = resolveSize(Animation.RELATIVE_TO_SELF, mFromRect.bottom, height, parentHeight).toInt()
        mResolvedTo.left = resolveSize(Animation.RELATIVE_TO_SELF, mToRect.left, width, parentWidth).toInt()
        mResolvedTo.top = resolveSize(Animation.RELATIVE_TO_SELF, mToRect.top, height, parentHeight).toInt()
        mResolvedTo.right = resolveSize(Animation.RELATIVE_TO_SELF, mToRect.right, width, parentWidth).toInt()
        mResolvedTo.bottom = resolveSize(Animation.RELATIVE_TO_SELF, mToRect.bottom, height, parentHeight).toInt()
    }

    override fun applyTransformation(it: Float, tr: Transformation) {
        val l = mResolvedFrom.left + ((mResolvedTo.left - mResolvedFrom.left) * it).toInt()
        val t = mResolvedFrom.top + ((mResolvedTo.top - mResolvedFrom.top) * it).toInt()
        val r = mResolvedFrom.right + ((mResolvedTo.right - mResolvedFrom.right) * it).toInt()
        val b = mResolvedFrom.bottom + ((mResolvedTo.bottom - mResolvedFrom.bottom) * it).toInt()

        try {
            mSetClipRect!!.invoke(tr, l, t, r, b)
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        }
    }

    override fun willChangeTransformationMatrix(): Boolean {
        return false
    }
}